pkg/avx/collector.go

```diff
@@ -70,118 +74,229 @@ var descriptors = [numDescriptors]*prometheus.Desc{
-func getElfKernelVersion(path string) (uint8, uint8, uint8, error) {
-	elfFile, err := elf.Open(path)
-	if err != nil {
-		return 0, 0, 0, errors.Wrapf(err, "unable to open ELF file %s", path)
-	}
-	defer elfFile.Close()
+func enablePerfTracepoint(fd int, tracepoint string) (int, error) {
 
-	sec := elfFile.Section("version")
-	if sec == nil {
-		return 0, 0, 0, errors.New("unable to find 'version' section")
-	}
-	data, err := sec.Data()
+	path := filepath.Join(kernelTracepointPath, tracepoint, "id")
+	log.Debug("tracepoint: %s", path)
+
+	id, err := ioutil.ReadFile(path)
 	if err != nil {
-		return 0, 0, 0, errors.Wrap(err, "unable to get version data")
+		return -1, errors.Wrap(err, "unable read tracepoint ID")
 	}
-
-	// Least Significant Byte first
-	return data[2], data[1], data[0], nil
-}
-
-func checkElfKernelVersion(path string) error {
-	elfMajor, elfMinor, _, err := getElfKernelVersion(path)
+	tid, err := strconv.Atoi(strings.TrimSpace(string(id)))
 	if err != nil {
-		return err
+		return -1, errors.New("unable to convert tracepoint ID")
 	}
 
-	currentCode, err := bpf.CurrentKernelVersion()
+	log.Debug("tracepoint ID: %d", tid)
+	attr := unix.PerfEventAttr{
+		Type:        unix.PERF_TYPE_TRACEPOINT,
+		Config:      uint64(tid), // tracepoint id
+		Sample_type: unix.PERF_SAMPLE_RAW,
+		Sample:      1,
+		Wakeup:      1,
+	}
+	pfd, err := unix.PerfEventOpen(&attr, -1, 0, -1, unix.PERF_FLAG_FD_CLOEXEC)
 	if err != nil {
-		return errors.Wrap(err, "unable to get current kernel version")
+		return -1, errors.Wrap(err, "unable open perf events")
 	}
-
-	if currentCode < kernelVersionCode(elfMajor, elfMinor, 0) {
-		return errors.New("host kernel is too old, consider rebuilding eBPF")
+	if _, _, errno := unix.Syscall(unix.SYS_IOCTL, uintptr(pfd), unix.PERF_EVENT_IOC_ENABLE, 0); errno != 0 {
+		return -1, errors.New("unable to set up perf events")
+	}
+	if _, _, errno := unix.Syscall(unix.SYS_IOCTL, uintptr(pfd), unix.PERF_EVENT_IOC_SET_BPF, uintptr(fd)); errno != 0 {
+		return -1, errors.New("unable to attach bpf program to perf events")
 	}
 
-	return nil
-}
-
-type collector struct {
-	root                     string
-	elfFilepath              string
-	bpfModule                *bpf.Module
-	avxContextSwitchCounters *bpf.Map
-	allContextSwitchCounters *bpf.Map
-	lastUpdateNs             *bpf.Map
-	lastCPUCounters          *bpf.Map
+	return pfd, nil
 }
 
 // NewCollector creates new Prometheus collector for AVX metrics
 func NewCollector() (prometheus.Collector, error) {
 
-	elfFilepath := path.Join(bpfInstallpath, bpfBinaryName)
-
-	if err := checkElfKernelVersion(elfFilepath); err != nil {
-		return nil, err
-	}
-
-	bpfModule := bpf.NewModule(elfFilepath)
-
-	sectionParams := make(map[string]bpf.SectionParams)
-	if err := bpfModule.Load(sectionParams); err != nil {
-		return nil, errors.Wrap(err, "unable to load eBPF ELF file")
+	// TODO: get rid of this
+	// Increase `ulimit -l` limit to avoid BPF_PROG_LOAD error (runc #2167).
```

Comment:
Could this be some Systemd directive such as `LimitMEMLOCK=`? Maybe just check that the limit is large enough and error out if it isn't.

Reply:
this piece is still TODO. I want to understand better why it's needed in the first place (we didn't need it earlier)